components/webext-storage/src/sync/mod.rs

```diff
@@ -0,0 +1,156 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+mod incoming;
+mod outgoing;
+
+#[cfg(test)]
+mod sync_tests;
+
+use serde_derive::*;
+use serde_json;
+use sync15_traits::ServerTimestamp;
+use sync_guid::Guid as SyncGuid;
+
+use incoming::IncomingAction;
+
+type JsonMap = serde_json::Map<String, serde_json::Value>;
+
+/// For use with `#[serde(skip_serializing_if = )]`
+#[inline]
+pub fn is_default<T: PartialEq + Default>(v: &T) -> bool {
+    *v == T::default()
+}
+
+// XXX - need to work out how to consolidate "payload" vs "bso", particularly
+// the timestamp, etc. For now, we represent this in ServerPayload, even though
+// that's never actually stored in this way.
+// XXX - this isn't going to work without other changes.
```

Comment:
Hmmm, yeah, we're going to need to pass that over in Desktop. I wonder if we can "just" pass the raw decrypted JSON over XPCOM, and have the Desktop integration layer parse it into a `sync15_traits::Payload` (which has `id`, `deleted`, and `data` already on it) before handing it over to `webext_storage`.

Reply:
I spent a bit of time noodling on this, and the result is #3067 and https://phabricator.services.mozilla.com/D73581. Basically, we have this thing called an "envelope", which is BSO metadata fields + the unparsed payload. An incoming envelope has a `payload()` method, which you can use to get at the payload—which you can then either deserialize into a record, or check if it's a tombstone via `is_deleted`, like you can in a-s `sync15`.

I think we can land your patch as-is now, and, assuming you like the envelope idea, go back and tweak this after.